mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -34,33 +35,134 @@
 namespace mitigation {
 
+bool containsQnodes(func::FuncOp funcOp)
+{
+    bool containsQnodes = false;
+    funcOp.walk([&](func::CallOp op) {
+        auto insideFuncOp =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+        if (insideFuncOp->hasAttr("qnode")) {
+            containsQnodes = true;
+            return WalkResult::interrupt();
+        }
+        return WalkResult::advance();
+    });
+    return containsQnodes;
+}
+
+func::FuncOp createZneFunc(func::FuncOp funcOp, PatternRewriter &rewriter)
+{
+    auto loc = funcOp.getLoc();
+    TypeRange originalTypes = funcOp.getArgumentTypes();
+    SmallVector<Type> typesFolded(originalTypes.begin(), originalTypes.end());
+    Type indexType = rewriter.getIndexType();
+    typesFolded.push_back(indexType);
+    FunctionType fnFoldedType = FunctionType::get(funcOp.getContext(),
+                                                  /*inputs=*/typesFolded,
+                                                  /*outputs=*/funcOp.getResultTypes());
+    std::string fnFoldedName = funcOp.getName().str() + ".zne";
+    rewriter.setInsertionPointToStart(funcOp->getParentOfType<ModuleOp>().getBody());
+    auto fnFoldedOp = rewriter.create<func::FuncOp>(loc, fnFoldedName, fnFoldedType);
+
+    rewriter.cloneRegionBefore(funcOp.getBody(), fnFoldedOp.getBody(), fnFoldedOp.end());
+
+    Block *fnFoldedOpBlock = &fnFoldedOp.getBody().front();
+    fnFoldedOpBlock->addArgument(fnFoldedOp.getArgumentTypes().back(), loc);
+    return fnFoldedOp;
+}
+
 LogicalResult ZneLowering::match(mitigation::ZneOp op) const { return success(); }
 
+// TODO: Optimize the traversal of call graphs (currently used twice)
+// Also all functions exploree in the call graph get their ZNE version.
 void ZneLowering::rewrite(mitigation::ZneOp op, PatternRewriter &rewriter) const
 {
     Location loc = op.getLoc();
-
+    auto moduleOp = op->getParentOfType<ModuleOp>();
     // Number of folds
     auto numFolds = op.getNumFolds();
     RankedTensorType numFoldType = cast<RankedTensorType>(numFolds.getType());
     const auto sizeInt = numFoldType.getDimSize(0);
 
     // Folding type
     auto foldingAlgorithm = op.getFolding();
-
-    // Create the folded circuit function
-    FlatSymbolRefAttr foldedCircuitRefAttr =
-        getOrInsertFoldedCircuit(loc, rewriter, op, foldingAlgorithm);
-    func::FuncOp foldedCircuit =
-        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, foldedCircuitRefAttr);
-
+    auto calleeOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+
+    func::FuncOp fnFoldedOp;
+
+    if (!calleeOp->hasAttr("qnode")) {
+        // Traverse the callgraph, copy all the function to a `.zne` version and fold qnodes
+        traverseCallGraph(calleeOp, /*symbolTable=*/nullptr, [&](func::FuncOp funcOp) {
+            if (!funcOp->hasAttr("qnode")) {
+                PatternRewriter::InsertionGuard insertGuard(rewriter);
```

Comment:
~~This guard appears to be unnecessary~~ or rather this is the missing rewrite guard in the `createZneFunc` op from my previous comment.

Reply:
Moved it :+1: